### PR TITLE
sql: Add all has_X_privilege functions

### LINKED ISSUE #19183: SQL: Implement access privilege inquire functions
### Feature request

PostgreSQL has various functions that let users query their current privileges easily. We should implement these same functions in Materialize.

The full list can be found here: https://www.postgresql.org/docs/15/functions-info.html in Table 9.67

### PR DESCRIPTION
**Pull request by jkosh44**
35b5cf18f54e4f31f4676abafa34b2c95d2cd791 implemented the has_table_privilege function. This commit implements the same function for all other applicable object types. Specifically:

 - Clusters
 - Connections
 - Databases
 - Schemas
 - Secrets
 - Types

These functions come from PostgreSQL and allow a user to query their privileges on a specific object.

Resolves #19183

### Motivation
This PR adds a known-desirable feature.

### Tips for reviewer
* 953 lines of the 1,136 line diff comes from tests.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - This release adds the following functions:
    - `has_cluster_privilege()`
    - `has_connection_privilege()`
    - `has_database_privilege()`
    - `has_schema_privilege()`
    - `has_secret_privilege()`
    - `has_table_privilege()`
    - `has_type_privilege()`

### CODE REVIEW

**Comment by jkosh44** [comment]
Blocked on https://github.com/MaterializeInc/materialize/pull/19530

File: src/sql/src/func.rs [hunk 1]
```diff
@@ -1889,36 +1931,37 @@ pub static PG_CATALOG_BUILTINS: Lazy<BTreeMap<&'static str, Func>> = Lazy::new(|
...
             params!(Bytes, Int32) => BinaryFunc::GetByte => Int32, 721;
         },
-        "has_table_privilege" => Scalar {
-            params!(String, String, String) => sql_impl_func("has_table_privilege(mz_internal.mz_role_oid($1), $2::regclass::oid, $3)") => Bool, 1922;
-            params!(String, Oid, String) => sql_impl_func("has_table_privilege(mz_internal.mz_role_oid($1), $2, $3)") => Bool, 1923;
-            params!(Oid, String, String) => sql_impl_func("has_table_privilege($1, $2::regclass::oid, $3)") => Bool, 1924;
-            params!(Oid, Oid, String) => sql_impl_func("
+        "has_cluster_privilege" => Scalar {
+            params!(String, String, String) => sql_impl_func("has_cluster_privilege(mz_internal.mz_role_oid($1), $2, $3)") => Bool, oid::FUNC_HAS_CLUSTER_PRIVILEGE_TEXT_TEXT_TEXT_OID;
+            params!(Oid, String, String) => sql_impl_func("
```
Comment (mjibson) [inline src/sql/src/func.rs:1936]: Why can't this one use the macro?

File: doc/user/data/sql_funcs.yml [hunk 2]
```diff
@@ -776,10 +776,44 @@
 
 - type: Access privilege inquiry
   description: |
-    Functions that allow querying object access privileges.
+    Functions that allow querying object access privileges. All of the following functions have
+    limitations involved in resolving the text names passed in. None of the object names can have
+    any schema or database qualifiers. For example, `has_table_privilege('public.t', 'SELECT')`
+    will not work properly. Additionally, the name lookup does not consider the current search path
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply doc/user/data/sql_funcs.yml:782]: The name lookup relies on the `text` to `regclass` casting, which is supposed to do name look-ups that allow qualified names, use the current search path, and use the active database. Instead that cast just does a global lookup in `mz_objects`.

The implementation is here: 
https://github.com/MaterializeInc/materialize/blob/84d2d3ce939e91987ecf3910704b42c220d88646/src/sql/src/plan/typeconv.rs#L390-L414

File: src/sql/src/func.rs (hunk 1 above)
Comment (jkosh44) [reply src/sql/src/func.rs:1936]: The macro relies on the object having an OID, and clusters don't have OIDs. I can add a comment to make this clear.

File: doc/user/data/sql_funcs.yml [hunk 3]
```diff
@@ -776,10 +776,44 @@
...
+    or active database.
   functions:
+    - signature: 'has_cluster_privilege([role: text or oid, ]cluster: text, privileges: text) -> bool'
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply doc/user/data/sql_funcs.yml]: Done

**Review by mjibson — APPROVED** [review]

**Pull request merged**